anonymizer/OpenDiffix.Service/Led.fs

```diff
@@ -0,0 +1,176 @@
+    CountArguments: Collections.Generic.List<Value list>
+    ExecutionContext: ExecutionContext
+    OutputRow: Row
+    mutable IsLowCount: bool
+  }
+
+let private mergeBucket groupingLabelsLength lowCountIndex countIndex destinationBucket fromBucket =
+  let lowCountAggregator = destinationBucket.Aggregators.[lowCountIndex]
+  let countAggregator = destinationBucket.Aggregators.[countIndex]
+
+  fromBucket.LowCountArguments |> Seq.iter lowCountAggregator.Transition
+  fromBucket.CountArguments |> Seq.iter countAggregator.Transition
+
+  destinationBucket.OutputRow.[groupingLabelsLength + lowCountIndex] <-
+    lowCountAggregator.Final destinationBucket.ExecutionContext
+
+  destinationBucket.OutputRow.[groupingLabelsLength + countIndex] <-
+    countAggregator.Final destinationBucket.ExecutionContext
+
+let private LOCAL_AGGREGATION = false
+
+let private executeLed executionContext (childPlan, groupingLabels, aggregators) : seq<Row> =
+  let groupingLabels = Array.ofList groupingLabels
+  let aggFns, aggArgs = aggregators |> Array.ofList |> Executor.Utils.unpackAggregators
+  let groupingLabelsLength = groupingLabels.Length
+  let rowSize = groupingLabelsLength + aggFns.Length
+
+  let lowCountIndex = findAggregator DiffixLowCount aggFns
+  let countIndex = findAggregator DiffixCount aggFns
+
+  let mergeBucketInto = mergeBucket groupingLabelsLength lowCountIndex countIndex
+
+  let makeBucket (group: Value array) =
+    let bucketSeed = Executor.Utils.addValuesToSeed executionContext.NoiseLayers.BucketSeed group
+
+    {
+      Aggregators = aggFns |> Array.map (Aggregator.create executionContext LOCAL_AGGREGATION)
+      LowCountArguments = Collections.Generic.List<Value list>()
+      CountArguments = Collections.Generic.List<Value list>()
+      ExecutionContext =
+        { executionContext with
+            NoiseLayers = { executionContext.NoiseLayers with BucketSeed = bucketSeed }
+        }
+      // Aggregator values are computed later
+      OutputRow = Array.init rowSize (fun i -> if i < groupingLabelsLength then group.[i] else Null)
+      IsLowCount = false
+    }
+
+  let state = Collections.Generic.Dictionary<Row, BucketState>(Row.equalityComparer)
+
+  // Group rows into buckets and keep track of low-count/count arguments
+  for row in Executor.execute executionContext childPlan do
+    let group = groupingLabels |> Array.map (Expression.evaluate row)
+    let argEvaluator = Expression.evaluate row
+
+    let bucket =
+      match state.TryGetValue(group) with
+      | true, bucketState -> bucketState
+      | false, _ ->
+        let bucketState = makeBucket group
+        state.[group] <- bucketState
+        bucketState
+
+    bucket.Aggregators
+    |> Array.iteri (fun i aggregator ->
+      let args = aggArgs.[i] |> List.map argEvaluator
+
+      if i = lowCountIndex then bucket.LowCountArguments.Add(args)
+      else if i = countIndex then bucket.CountArguments.Add(args)
```

Comment:
It would be more efficient here (but also more complex overall) to instead provide a generic way in `reference` to merge internal aggregator states.

Reply:
You mean an `IAggregator` should support a `.Merge` function?

Reply:
Yes.

Reply:
With that we can afterwards make a generic `merge` function for buckets.
And instead of having a full `Aggregation` hook, we could have a lighter `PostAggregation` one.

Reply:
With the funny internal aggregator state we hold I was hesitant to fiddle there. LowCount appears easy enough, as for count and count distinct we'll have to look into it more carefully.